packages/clay-dropdown/src/ClayDropdownBase.js

```diff
@@ -123,6 +123,180 @@ class ClayDropdownBase extends ClayComponent {
+
+				active = nextActive < 0 ? totalItems - 1 : nextActive;
+
+				subActive = items[active].length - 1;
+			}
+		} else {
+			if (subActive === items[active].length) {
+				const nextActive = active + 1;
+
+				active = nextActive === totalItems ? 0 : nextActive;
+
+				subActive = 0;
+			}
+		}
+
+		return {active, subActive};
+	}
+
+	/**
+	 * Gets the indexes of the next active item.
+	 * @param {Array} items
+	 * @param {Number} initialActive
+	 * @param {Number} initialSubActive
+	 * @param {Boolean} reverse
+	 * @private
+	 * @return {Number}
+	 */
+	_getNextIndexes(items, initialActive, initialSubActive, reverse = false) {
+		let {active, subActive} = this._incrementActiveItemIndex(
+			items,
+			initialActive === -1 ? 0 : initialActive,
+			initialSubActive,
+			reverse
+		);
+
+		const totalItems = items.flat(2);
+
+		let tick = 0;
+
+		while (items[active][subActive].disabled && tick < totalItems.length) {
+			const item = this._incrementActiveItemIndex(
+				items,
+				active,
+				subActive,
+				reverse
+			);
+
+			active = item.active;
+			subActive = item.subActive;
+
+			tick++;
+		}
+
+		return {active, subActive};
+	}
+
+	/**
+	 * Sets the next item in the list as active.
+	 * @param {Boolean} reverse
+	 * @private
+	 */
+	_setNextActive(reverse) {
+		const items = this._flattenItems();
+		let activeSubIndex = -1;
+
+		const activeIndex = items.findIndex(item => {
+			const subIndex = item.findIndex(item => item.active);
+
+			if (subIndex !== -1) {
+				activeSubIndex = subIndex;
+
+				return true;
+			}
+		});
+
+		const {active, subActive} = this._getNextIndexes(
+			items,
+			activeIndex,
+			activeSubIndex,
+			reverse
+		);
+
+		if (items[activeIndex] && items[activeIndex][activeSubIndex]) {
+			items[activeIndex][activeSubIndex].active = false;
+		}
+
+		items[active][subActive].active = true;
+
+		this.items = this._assembleFromFlattenedItems(items);
+	}
+
+	/**
+	 * Flattens all items into an array of arrays.
+	 * @private
+	 * @return {Array}
+	 */
+	_flattenItems() {
+		return this.items.map(item => (item.items ? item.items : [item]));
+	}
+
+	/**
+	 * Reassembles flattened items into the correct items structure.
+	 * @param {Array} flattenedItems
+	 * @private
+	 * @return {Array}
+	 */
+	_assembleFromFlattenedItems(flattenedItems) {
+		return this.items.map((item, i) => {
+			if (item.items) {
+				item.items = flattenedItems[i];
+			} else {
+				item = flattenedItems[i][0];
+			}
+
+			return item;
+		});
+	}
+
+	/**
+	 * Handle key events.
+	 * @param {!Event} event
+	 * @private
+	 */
+	_handleKeyNavigation(event) {
+		event.preventDefault();
+
+		if (event.key === 'ArrowDown' || event.key === 'ArrowUp') {
+			this._setNextActive(event.key === 'ArrowUp');
+		} else if (event.key === 'Enter') {
+			this._handleCheckedToggle();
+		}
+	}
+
+	/**
+	 * Toggles the 'checked' attribute on the active item.
+	 * @private
+	 */
+	_handleCheckedToggle() {
+		const flattenedItems = this._flattenItems();
```

Comment:
hey @bryceosterhaus what do you think of adding a `memoize` here, decreases the cost of flattening the array of items every time the function is called during user interaction, we just have to make sure to clean it whenever the array of items is changed (we can guarantee this by passing `this.items` as parameter).